gaas/src/main/kotlin/tw/waterballsa/utopia/gaas/AbsenceRecord.kt

```diff
@@ -0,0 +1,135 @@
+
+fun removeCanceledOrCompletedEvent() = listener {
+    on<ScheduledEventUpdateStatusEvent> {
+        scheduledEvent
+            .takeIf { it.isStudyCircleEvent() && (it.status == ScheduledEvent.Status.CANCELED || it.status == ScheduledEvent.Status.COMPLETED) }
+            ?.run {
+                log.info { "Remove cancel or completed event, event id: $id" }
+                studyCircleEventIds.remove(id)
+            }
+    }
+}
+
+//活動開始時，判斷是否是 GaaS 讀書會，是則啟動 task，每隔一段時間檢查參加者名單並記錄
+fun recordTask(event: ScheduledEvent) {
+    val filePath = createDataFile()
+
+    val now = getTaipeiCurrentDateTime()
+    val start = now.withHour(21).withMinute(0).withSecond(0)
+    val end = now.withHour(22).withMinute(0).withSecond(0)
+    val participantCount = mutableListOf<Int>()
+
+    timer.schedule(object : TimerTask() {
+        override fun run() {
+            val eventChannel = event.channel
+            checkNotNull(eventChannel)
+
+            eventChannel.asVoiceChannel().run {
+                participantCount.add(members.size)
+                val newContent = members.map { "${it.nickname} : ${it.id}" }
+                val currentTime = dateFormatter.format(getTaipeiCurrentDateTime())
+                writeParticipantsIntoFile(currentTime, newContent, filePath)
+            }
+        }
+    }, getTaipeiTime(start), 3.minutes.inWholeMilliseconds)
+
+    timer.schedule(
+        object : TimerTask() {
+            override fun run() {
+                writeStaticsSummaryIntoFile(participantCount, filePath)
+                timer.cancel()
+                log.info { "Record Task has been closed." }
+            }
+        },
+        getTaipeiTime(end)
+    )
+}
+
+@Synchronized
+private fun writeStaticsSummaryIntoFile(participantCount: List<Int>, filePath: Path) {
+    val avgStatics = "Avg: ${participantCount.average().roundToInt()}"
+    val maxStatics = "Max: ${participantCount.max()}"
+    writeString(filePath, avgStatics + lineSeparator(), APPEND)
+    writeString(filePath, maxStatics + lineSeparator(), APPEND)
+}
+
+@Synchronized
+private fun writeParticipantsIntoFile(currentTime: String, newContent: List<String>, filePath: Path) {
+    writeString(filePath, currentTime + lineSeparator(), APPEND)
+    newContent.forEach { writeString(filePath, it + lineSeparator(), APPEND) }
+    writeString(filePath, lineSeparator(), APPEND)
+}
+
+private fun createDataFile(): Path {
+    val fileName = DATAFILE_NAME_TEMPLATE.replace("\$date", getTaipeiCurrentDate().toString())
+    val filePath = Paths.get(DATABASE_PATH + fileName)
+    val file = filePath.toFile()
+
+    if (!file.exists()) {
+        log.info { "Data File Created: ${file.name}" }
+        createDirectories(Paths.get(DATABASE_PATH))
+        createFile(filePath)
+    }
+    return filePath
+}
+
+private fun getTaipeiCurrentDateTime(): LocalDateTime = LocalDateTime.now(ZoneId.of(TIMEZONE_ID))
+
+private fun getTaipeiCurrentDate(): LocalDate = LocalDate.now(ZoneId.of(TIMEZONE_ID))
```

Comment:
do we really have to specify the time zone in every feature?

I think we can set the default time zone in the main function to prevent duplicate time zone configs